File: packages/core/src/generator.ts
/**
 * SmartPayloadGenerator — generates payloads based on JSON Schema
 * and the selected fuzzing profile (RANDOM / BOUNDARY / MALICIOUS).
 *
 * BOUNDARY and MALICIOUS profiles use sequential iteration through their
 * payload arrays to guarantee full coverage. RANDOM uses true randomness.
 */

import type { Dictionary, FuzzingProfile, SchemaProperty } from './types.js';
import * as rand from './random.js';
import {
    BOUNDARY_STRINGS,
    BOUNDARY_INTEGERS,
    BOUNDARY_NUMBERS,
    BOUNDARY_DATES,
    BOUNDARY_BOOLEANS,
    BOUNDARY_ARRAY_SIZES,
} from './payloads/boundary.js';
import {
    ALL_MALICIOUS_STRINGS,
    MALICIOUS_NUMBERS,
    MALICIOUS_DATES,
    MALICIOUS_BOOLEANS,
    MALICIOUS_TYPE_CONFUSION,
} from './payloads/malicious.js';

// ─── Sequential counter key types ──────────────────────────

type BoundaryCounter = '_bStrIdx' | '_bIntIdx' | '_bNumIdx' | '_bDateIdx' | '_bArrIdx' | '_bBoolIdx';
type MaliciousCounter = '_mStrIdx' | '_mNumIdx' | '_mDateIdx' | '_mBoolIdx' | '_mTypeIdx';
type SeqCounter = BoundaryCounter | MaliciousCounter;

export class SmartPayloadGenerator {
    private dictionaries: Dictionary;
    private profile: FuzzingProfile;

    // ─── Sequential counters: BOUNDARY ──────────────────────
    private _bStrIdx = 0;
    private _bIntIdx = 0;
    private _bNumIdx = 0;
    private _bDateIdx = 0;
    private _bArrIdx = 0;
    private _bBoolIdx = 0;

    // ─── Sequential counters: MALICIOUS ─────────────────────
    private _mStrIdx = 0;
    private _mNumIdx = 0;
    private _mDateIdx = 0;
    private _mBoolIdx = 0;
    private _mTypeIdx = 0;

    constructor(dictionaries: Dictionary = {}, profile: FuzzingProfile = 'RANDOM') {
        // Normalize dictionary keys to lowercase
        this.dictionaries = Object.keys(dictionaries).reduce((acc, key) => {
            acc[key.toLowerCase()] = dictionaries[key];
            return acc;
        }, {} as Dictionary);
        this.profile = profile;
    }

    // ─── Static helpers ────────────────────────────────────

    /**
     * Minimum iterations needed to cover every payload in the given profile.
     * Returns 0 for RANDOM (no fixed set to exhaust).
     */
    public static minIterationsNeeded(profile: FuzzingProfile): number {
        switch (profile) {
            case 'BOUNDARY':
                return Math.max(
                    BOUNDARY_STRINGS.length,
                    BOUNDARY_INTEGERS.length,
                    BOUNDARY_NUMBERS.length,
                    BOUNDARY_DATES.length,
                    BOUNDARY_BOOLEANS.length,
                    BOUNDARY_ARRAY_SIZES.length,
                );
            case 'MALICIOUS':
                return Math.max(
                    ALL_MALICIOUS_STRINGS.length,
                    MALICIOUS_NUMBERS.length,
                    MALICIOUS_DATES.length,
                    MALICIOUS_BOOLEANS.length,
                    MALICIOUS_TYPE_CONFUSION.length,
                );
            default:
                return 0;
        }
    }

    /** @deprecated Use minIterationsNeeded('BOUNDARY') instead. */
    public static boundaryIterationsNeeded(): number {
        return SmartPayloadGenerator.minIterationsNeeded('BOUNDARY');
    }

    // ─── Sequential pick ───────────────────────────────────

    /** Advances the counter each call, wraps around at array length. */
    private seqPick<T>(arr: readonly T[], counter: SeqCounter): T {
        const val = arr[this[counter] % arr.length];
        (this as any)[counter]++;
        return val;
    }

    // ─── Public API ────────────────────────────────────────

    /**
     * Generate a value for a single property.
     * Priority: enum → dictionary → format-aware → profile-based
     */
    public generate(propertyName: string, schema: SchemaProperty): any {
        // 1. Enum — respect explicit enum values, but allow bypass in security profiles
        if (schema.enum && schema.enum.length > 0) {
            const shouldBypass =
                (this.profile === 'MALICIOUS' || this.profile === 'BOUNDARY') &&
                rand.next() < 0.3;
            if (!shouldBypass) {
                return rand.pick(schema.enum);
            }
            // Fall through to standard profile-based generation to probe out-of-spec values
        }

        // 2. User dictionary (highest priority after enum)
        const normalizedName = propertyName.toLowerCase();
        if (this.dictionaries[normalizedName] && this.dictionaries[normalizedName].length > 0) {
            return rand.pick(this.dictionaries[normalizedName]);
        }

        // 3. Profile-based generation
        return this.generateByProfile(schema.type, schema.format, propertyName);
    }

    /**
     * Recursively build a full object from JSON Schema.
     */
    public buildObject(schema: SchemaProperty): Record<string, any> {
        if (schema.type !== 'object' || !schema.properties) return {};

        const payload: Record<string, any> = {};

        for (const [key, propSchema] of Object.entries(schema.properties)) {
            const isRequired = schema.required?.includes(key) ?? false;

            // 30% chance to omit optional fields in intensive profiles
            if (
                !isRequired &&
                (this.profile === 'BOUNDARY' || this.profile === 'MALICIOUS') &&
                rand.next() < 0.3
            ) {
                continue;
            }

            // 5% chance to omit REQUIRED fields in MALICIOUS profile to test server validation
            if (isRequired && this.profile === 'MALICIOUS' && rand.next() < 0.05) {
                continue;
            }

            if (propSchema.type === 'object' && propSchema.properties) {
                payload[key] = this.buildObject(propSchema);
            } else if (propSchema.type === 'array' && propSchema.items) {
                const count = this.getArraySize(propSchema.items);
                payload[key] = Array.from({ length: count }, () =>
                    propSchema.items!.type === 'object'
                        ? this.buildObject(propSchema.items!)
                        : this.generate(key, propSchema.items!),
                );
            } else {
                payload[key] = this.generate(key, propSchema);
            }
        }

        return payload;
    }

    // ─── Private ────────────────────────────────────────────

    private getArraySize(itemSchema?: SchemaProperty): number {
        if (this.profile === 'BOUNDARY') {
            const size = this.seqPick(BOUNDARY_ARRAY_SIZES, '_bArrIdx');
            // Cap complex object arrays to prevent OOM; primitives can be huge
            return itemSchema?.type === 'object' ? Math.min(size, 50) : size;
        }
        return rand.int(1, 5);
    }

    private generateByProfile(type?: string, format?: string, propName?: string): any {
        // MALICIOUS: 5% chance to completely break the expected type
        if (this.profile === 'MALICIOUS' && rand.next() < 0.05) {
            return this.breakType();
        }

        // Handle format-specific date-time before generic string
        if (type === 'string' && format === 'date-time') {
            return this.generateDate();
        }

        switch (type) {
            case 'string':
                return this.generateString(format, propName);
            case 'integer':
            case 'number':
                return this.generateNumber(type);
            case 'boolean':
                return this.generateBoolean();
            default:
                // Fallback for unknown type — try to guess by name
                if (propName) {
                    const lower = propName.toLowerCase();
                    if (lower.includes('id') || lower.includes('uuid')) return rand.uuid();
                    if (lower.includes('slug') || lower.includes('name')) return rand.word();
                    if (lower.includes('num') || lower.includes('count') || lower.includes('page')) return rand.int(1, 100);
                }
                return rand.randomString(rand.int(3, 10));
        }
    }

    private generateString(format?: string, propName?: string): any {
        // If it's a generic string without format, check name-based heuristics first
        if (!format && propName) {
            const lower = propName.toLowerCase();
            if (lower.includes('id') || lower.includes('uuid')) return rand.uuid();
            if (lower.includes('slug') || lower.includes('name')) return rand.word();
            if (lower.includes('num') || lower.includes('count') || lower.includes('page')) return String(rand.int(1, 100));
        }

        switch (this.profile) {
            case 'BOUNDARY':
                return this.seqPick(BOUNDARY_STRINGS, '_bStrIdx');

            case 'MALICIOUS':
                return this.seqPick(ALL_MALICIOUS_STRINGS, '_mStrIdx');

            case 'RANDOM':
            default:
                return this.generateRandomString(format);
        }
    }

    private generateRandomString(format?: string): string {
        switch (format) {
            case 'uuid':
                return rand.uuid();
            case 'email':
                return rand.email();
            case 'uri':
            case 'url':
                return rand.uri();
            case 'ipv4':
            case 'ip':
                return rand.ipv4();
            case 'date-time':
                return rand.date().toISOString();
            default:
                return rand.word();
        }
    }

    private generateNumber(type?: string): any {
        switch (this.profile) {
            case 'BOUNDARY':
                return type === 'integer'
                    ? this.seqPick(BOUNDARY_INTEGERS, '_bIntIdx')
                    : this.seqPick([...BOUNDARY_INTEGERS, ...BOUNDARY_NUMBERS], '_bNumIdx');

            case 'MALICIOUS':
                return this.seqPick(MALICIOUS_NUMBERS, '_mNumIdx');

            case 'RANDOM':
            default:
                return type === 'integer'
                    ? rand.int(1, 1000)
                    : rand.float(0, 1000);
        }
    }

    private generateBoolean(): any {
        switch (this.profile) {
            case 'BOUNDARY':
                return this.seqPick(BOUNDARY_BOOLEANS, '_bBoolIdx');

            case 'MALICIOUS':
                return this.seqPick(MALICIOUS_BOOLEANS, '_mBoolIdx');

            case 'RANDOM':
            default:
                return rand.bool();
        }
    }

    private generateDate(): any {
        switch (this.profile) {
            case 'BOUNDARY':
                return this.seqPick(BOUNDARY_DATES, '_bDateIdx');

            case 'MALICIOUS':
                return this.seqPick(MALICIOUS_DATES, '_mDateIdx');

            case 'RANDOM':
            default:
                return rand.date().toISOString();
        }
    }

    /** MALICIOUS: intentionally return wrong type to test strict typing */
    private breakType(): any {
        return this.seqPick(MALICIOUS_TYPE_CONFUSION, '_mTypeIdx');
    }
}

File: packages/container/internal/generator/generator.go
// Copyright (c) 2026 Swazz Authors
// This file is part of Swazz
// Swazz is licensed under the Business Source License 1.1 (BSL 1.1)
// See the LICENSE file in the project root or visit https://github.com/SecH0us3/swazz for more details

package generator

import (
	"fmt"
	"math/rand/v2"
	"strings"
	"sync"

	"github.com/google/uuid"

	"swazz-engine/internal/generator/payloads"
	"swazz-engine/internal/oob"
	"swazz-engine/internal/swagger"
)

// Generator produces fuzz payloads based on JSON Schema and a fuzzing profile.
type Generator struct {
	dictionaries map[string][]any
	profile      swagger.FuzzingProfile
	// activeCategories stores set of enabled category IDs for the current profile.
	// If nil, all categories are enabled.
	activeCategories map[string]bool

	// Sequential counters: BOUNDARY
	bStrIdx, bIntIdx, bNumIdx, bDateIdx, bArrIdx, bBoolIdx, bUUIDIdx int

	// Sequential counters: MALICIOUS
	mStrIdx, mNumIdx, mDateIdx, mBoolIdx, mTypeIdx, mUUIDIdx, oobIdx int

	// Sequential counter: security header rotation
	mu             sync.Mutex
	mSecHeaderIdxs map[string]int

	// cachedMaliciousStrings avoids allocations under high concurrency
	cachedMaliciousStrings []any
	// hasActiveMaliciousStrings tracks if any malicious string categories are enabled
	hasActiveMaliciousStrings bool

	oobServerURL string
	Endpoint     string
	RunID        string
	settings     swagger.Settings
}

// New creates a new Generator.
func New(dictionaries map[string][]any, profile swagger.FuzzingProfile, settings swagger.Settings) *Generator {
	norm := make(map[string][]any, len(dictionaries))
	for k, v := range dictionaries {
		norm[strings.ToLower(k)] = v
	}

	var active map[string]bool
	if settings.PayloadCategories != nil {
		if ids, ok := settings.PayloadCategories[profile]; ok && len(ids) > 0 {
			active = make(map[string]bool, len(ids))
			for _, id := range ids {
				active[id] = true
			}
		}
	}

	g := &Generator{
		dictionaries:     norm,
		profile:          profile,
		activeCategories: active,
		settings:         settings,
		oobServerURL:     settings.OOBServerURL,
		mSecHeaderIdxs:   make(map[string]int),
	}
	g.cachedMaliciousStrings, g.hasActiveMaliciousStrings = g.getActiveMaliciousStrings()
	return g
}

func (g *Generator) isCategoryEnabled(id string) bool {
	if g.activeCategories == nil {
		return true
	}
	return g.activeCategories[id]
}

func seqPick[T any](mu sync.Locker, arr []T, counter *int) T {
	mu.Lock()
	defer mu.Unlock()
	if len(arr) == 0 {
		var zero T
		return zero
	}
	val := arr[*counter%len(arr)]
	*counter++
	return val
}

func maxInt(nums ...int) int {
	m := 0
	for _, n := range nums {
		if n > m {
			m = n
		}
	}
	return m
}

// MinIterationsNeeded returns the minimum iterations required to cover all active payloads in a profile.
func MinIterationsNeeded(profile swagger.FuzzingProfile, settings swagger.Settings) int {
	var active map[string]bool
	if settings.PayloadCategories != nil {
		if ids, ok := settings.PayloadCategories[profile]; ok && len(ids) > 0 {
			active = make(map[string]bool, len(ids))
			for _, id := range ids {
				active[id] = true
			}
		}
	}

	is := func(id string) bool {
		if active == nil {
			return true
		}
		return active[id]
	}

	switch profile {
	case swagger.ProfileBoundary:
		max := 0
		if is(payloads.CatBoundaryStrings) && len(payloads.BoundaryStrings) > max {
			max = len(payloads.BoundaryStrings)
		}
		if is(payloads.CatBoundaryIntegers) && len(payloads.BoundaryIntegers) > max {
			max = len(payloads.BoundaryIntegers)
		}
		if is(payloads.CatBoundaryNumbers) && len(payloads.BoundaryNumbers) > max {
			max = len(payloads.BoundaryNumbers)
		}
		if is(payloads.CatBoundaryDates) && len(payloads.BoundaryDates) > max {
			max = len(payloads.BoundaryDates)
		}
		if is(payloads.CatBoundaryBooleans) && len(payloads.BoundaryBooleans) > max {
			max = len(payloads.BoundaryBooleans)
		}
		if is(payloads.CatBoundaryArrays) && len(payloads.BoundaryArraySizes) > max {
			max = len(payloads.BoundaryArraySizes)
		}
		if is(payloads.CatBoundaryUUIDs) && len(payloads.BoundaryUUIDs) > max {
			max = len(payloads.BoundaryUUIDs)
		}
		return max

	case swagger.ProfileMalicious:
		bodyCount := 0
		var maliciousBody []any
		if is(payloads.CatMaliciousEncoding) {
			maliciousBody = append(maliciousBody, payloads.MaliciousEncoding...)
		}
		if is(payloads.CatMaliciousSQLi) {
			maliciousBody = append(maliciousBody, payloads.MaliciousSQLi...)
		}
		if is(payloads.CatMaliciousXSS) {
			maliciousBody = append(maliciousBody, payloads.MaliciousXSS...)
		}
		if is(payloads.CatMaliciousPathTraversal) {
			maliciousBody = append(maliciousBody, payloads.MaliciousPathTraversal...)
		}
		if is(payloads.CatOOBInteraction) {
			maliciousBody = append(maliciousBody, payloads.MaliciousOOB...)
		}
		if is(payloads.CatMaliciousCmdi) {
			maliciousBody = append(maliciousBody, payloads.MaliciousCmdi...)
		}
		if is(payloads.CatMaliciousSSTI) {
			maliciousBody = append(maliciousBody, payloads.MaliciousSSTI...)
		}
		if is(payloads.CatMaliciousXXE) {
			maliciousBody = append(maliciousBody, payloads.MaliciousXXE...)
		}
		bodyCount = len(maliciousBody)
		if is(payloads.CatMaliciousNumbers) && len(payloads.MaliciousNumbers) > bodyCount {
			bodyCount = len(payloads.MaliciousNumbers)
		}
		if is(payloads.CatMaliciousDates) && len(payloads.MaliciousDates) > bodyCount {
			bodyCount = len(payloads.MaliciousDates)
		}
		if is(payloads.CatMaliciousBooleans) && len(payloads.MaliciousBooleans) > bodyCount {
			bodyCount = len(payloads.MaliciousBooleans)
		}
		if is(payloads.CatMaliciousTypeConfusion) && len(payloads.MaliciousTypeConfusion) > bodyCount {
			bodyCount = len(payloads.MaliciousTypeConfusion)
		}

		secHeaderCount := 0
		for _, def := range payloads.SecurityHeaderPayloads {
			if !is(def.Category) {
				continue
			}
			for _, values := range def.Headers {
				if len(values) > secHeaderCount {
					secHeaderCount = len(values)
				}
			}
		}

		return bodyCount + secHeaderCount
	default:
		return 0
	}
}

// Generate produces a value for a single property.
// Priority: enum → dictionary → format-aware → profile-based.
func (g *Generator) Generate(propertyName string, schema *swagger.SchemaProperty) any {
	if schema == nil {
		return nil
	}

	// 1. Enum — respect explicit values, allow bypass in security profiles
	if len(schema.Enum) > 0 {
		shouldBypass := (g.profile == swagger.ProfileMalicious) && rand.Float64() < 0.3 // #nosec G404 -- non-security randomness for fuzzer
		if !shouldBypass {
			return payloads.Pick(schema.Enum)
		}
	}

	// 2. User dictionary
	normalizedName := strings.ToLower(propertyName)
	if vals, ok := g.dictionaries[normalizedName]; ok && len(vals) > 0 {
		return payloads.Pick(vals)
	}

	// 3. Profile-based generation
	return g.generateByProfile(schema.Type, schema.Format, propertyName)
}

// BuildObject recursively builds a full object from JSON Schema.
func (g *Generator) BuildObject(schema *swagger.SchemaProperty) map[string]any {
	return g.buildObjectWithDepth(schema, 0)
}

func (g *Generator) buildObjectWithDepth(schema *swagger.SchemaProperty, depth int) map[string]any {
	if schema == nil || schema.Type != "object" || schema.Properties == nil || depth > 6 {
		return map[string]any{}
	}

	if g.profile == swagger.ProfileRandom && rand.Float64() < 0.1 { // #nosec G404
		return map[string]any{}
	}

	payload := make(map[string]any, len(schema.Properties))

	for key, propSchema := range schema.Properties {
		if propSchema == nil {
			continue
		}
		isRequired := false
		for _, r := range schema.Required {
			if r == key {
				isRequired = true
				break
			}
		}

		// 30% chance to omit optional fields in MALICIOUS profile
		if !isRequired &&
			(g.profile == swagger.ProfileMalicious) &&
			rand.Float64() < 0.3 { // #nosec G404 -- non-security randomness for fuzzer
			continue
		}

		// 5% chance to omit REQUIRED fields in MALICIOUS profile
		if isRequired && g.profile == swagger.ProfileMalicious && rand.Float64() < 0.05 { // #nosec G404 -- non-security randomness for fuzzer
			continue
		}

		if propSchema.Type == "object" && propSchema.Properties != nil {
			if depth < 6 {
				payload[key] = g.buildObjectWithDepth(propSchema, depth+1)
			} else {
				payload[key] = map[string]any{}
			}
		} else if propSchema.Type == "array" && propSchema.Items != nil {
			count := g.getArraySize(propSchema.Items)
			arr := make([]any, count)
			for i := range arr {
				if propSchema.Items.Type == "object" {
					if depth < 6 {
						arr[i] = g.buildObjectWithDepth(propSchema.Items, depth+1)
					} else {
						arr[i] = map[string]any{}
					}
				} else {
					arr[i] = g.GenerateArrayItem(key, propSchema.Items)
				}
			}
			payload[key] = arr
		} else {
			payload[key] = g.Generate(key, propSchema)
		}
	}

	return payload
}

// GenerateArrayItem produces a value for an array item, with safety caps to prevent OOM when multiplying large boundary strings across array lengths.
func (g *Generator) GenerateArrayItem(propertyName string, schema *swagger.SchemaProperty) any {
	if schema == nil {
		return nil
	}
	if schema.Type == "string" {
		formatLower := strings.ToLower(schema.Format)
		if formatLower == "date-time" {
			return g.generateDate()
		}
		if formatLower == "uuid" {
			return g.generateUUID()
		}
		return g.generateStringArrayItem(formatLower, propertyName)
	}
	return g.Generate(propertyName, schema)
}

// GenerateSecurityHeaders returns a map of header name → fuzz value for
// security-critical HTTP headers not defined in the API spec.
// Only active during MALICIOUS profile. Returns nil for other profiles.
func (g *Generator) GenerateSecurityHeaders() map[string]string {
	if g.profile != swagger.ProfileMalicious {
		return nil
	}

	g.mu.Lock()
	defer g.mu.Unlock()

	if g.mSecHeaderIdxs == nil {
		g.mSecHeaderIdxs = make(map[string]int)
	}

	headers := make(map[string]string)
	for _, def := range payloads.SecurityHeaderPayloads {
		if !g.isCategoryEnabled(def.Category) {
			continue
		}
		for headerName, values := range def.Headers {
			if len(values) == 0 {
				continue
			}
			idx := g.mSecHeaderIdxs[headerName]
			val := values[idx%len(values)]

			if strings.Contains(val, "{{OOB_URL}}") {
				u := uuid.New().String()
				url := g.oobURL(u)

				endpoint := "Header: " + headerName
				if g.Endpoint != "" {
					endpoint = fmt.Sprintf("%s (Header: %s)", g.Endpoint, headerName)
				}
				oob.GlobalStore.RegisterUUID(u, &oob.OOBContext{
					SessionID: g.RunID,
					Endpoint:  endpoint,
					Payload:   val,
				})
				val = strings.ReplaceAll(val, "{{OOB_URL}}", url)
			}

			headers[headerName] = val
			g.mSecHeaderIdxs[headerName] = idx + 1
		}
	}

	if len(headers) == 0 {
		return nil
	}
	return headers
}

// BodyIterations returns the number of iterations needed for body fuzzing.
func (g *Generator) BodyIterations() int {
	if g.profile != swagger.ProfileMalicious {
		return 0
	}
	is := g.isCategoryEnabled
	bodyCount := 0
	var maliciousBody []any
	if is(payloads.CatMaliciousEncoding) {
		maliciousBody = append(maliciousBody, payloads.MaliciousEncoding...)
	}
	if is(payloads.CatMaliciousSQLi) {
		maliciousBody = append(maliciousBody, payloads.MaliciousSQLi...)
	}
	if is(payloads.CatMaliciousXSS) {
		maliciousBody = append(maliciousBody, payloads.MaliciousXSS...)
	}
	if is(payloads.CatMaliciousPathTraversal) {
		maliciousBody = append(maliciousBody, payloads.MaliciousPathTraversal...)
	}
	if is(payloads.CatMaliciousCmdi) {
		maliciousBody = append(maliciousBody, payloads.MaliciousCmdi...)
	}
	if is(payloads.CatMaliciousSSTI) {
		maliciousBody = append(maliciousBody, payloads.MaliciousSSTI...)
	}
	if is(payloads.CatMaliciousXXE) {
		maliciousBody = append(maliciousBody, payloads.MaliciousXXE...)
	}
	if is(payloads.CatMaliciousPrototypePollution) {
		maliciousBody = append(maliciousBody, payloads.MaliciousPrototypePollution...)
	}
	if is(payloads.CatMaliciousNoSQLi) {
		maliciousBody = append(maliciousBody, payloads.MaliciousNoSQLi...)
	}
	if is(payloads.CatMaliciousSSRF) {
		maliciousBody = append(maliciousBody, payloads.MaliciousSSRF...)
	}
	if is(payloads.CatMaliciousMassAssignment) {
		maliciousBody = append(maliciousBody, payloads.MaliciousMassAssignment...)
	}
	if is(payloads.CatMaliciousGraphQL) {
		maliciousBody = append(maliciousBody, payloads.MaliciousGraphQL...)
	}
	if is(payloads.CatOOBInteraction) {
		maliciousBody = append(maliciousBody, payloads.MaliciousOOB...)
	}
	bodyCount = len(maliciousBody)
	if is(payloads.CatMaliciousNumbers) && len(payloads.MaliciousNumbers) > bodyCount {
		bodyCount = len(payloads.MaliciousNumbers)
	}
	if is(payloads.CatMaliciousDates) && len(payloads.MaliciousDates) > bodyCount {
		bodyCount = len(payloads.MaliciousDates)
	}
	if is(payloads.CatMaliciousBooleans) && len(payloads.MaliciousBooleans) > bodyCount {
		bodyCount = len(payloads.MaliciousBooleans)
	}
	if is(payloads.CatMaliciousTypeConfusion) && len(payloads.MaliciousTypeConfusion) > bodyCount {
		bodyCount = len(payloads.MaliciousTypeConfusion)
	}
	return bodyCount
}

// SecurityHeaderIterations returns the number of iterations needed for security header fuzzing.
func (g *Generator) SecurityHeaderIterations() int {
	if g.profile != swagger.ProfileMalicious {
		return 0
	}
	secHeaderCount := 0
	is := g.isCategoryEnabled
	for _, def := range payloads.SecurityHeaderPayloads {
		if !is(def.Category) {
			continue
		}
		for _, values := range def.Headers {
			if len(values) > secHeaderCount {
				secHeaderCount = len(values)
			}
		}
	}
	return secHeaderCount
}
